src/network/http/websocket.cpp

```diff
@@ -625,11 +627,35 @@ namespace fc { namespace http {
    }
    void websocket_server::listen( const fc::ip::endpoint& ep )
    {
-      my->_server.listen( boost::asio::ip::tcp::endpoint( boost::asio::ip::address_v4(uint32_t(ep.get_address())),ep.port()) );
+       my->_server.listen( boost::asio::ip::tcp::endpoint( boost::asio::ip::address_v4(uint32_t(ep.get_address())),ep.port()) );
+   }
+
+   uint16_t websocket_server::get_listening_port()
+   {
+       websocketpp::lib::asio::error_code ec;
+       return my->_server.get_local_endpoint(ec).port();
    }
 
    void websocket_server::start_accept() {
-      my->_server.start_accept();
+       my->_server.start_accept();
+   }
+
+   void websocket_server::stop_listening()
+   {
+       my->_server.stop_listening();
+   }
+
+   void websocket_server::close()
+   {
+       for (auto& connection : my->_connections)
+           my->_server.close(connection.first, websocketpp::close::status::normal, "Goodbye");
+   }
+
+   void websocket_server::synchronous_close()
+   {
+      close();
+      while (!my->_connections.empty())
+          fc::yield();
```

Comment:
Busy waiting is not nice. Can you replicate the `_closed` promise as used in the client?

Reply:
Well, it's not a busy wait because it yields, but you're still right: I can do better.

Reply:
Turns out there was already a `_closed` promise and I just didn't see it.